### PR TITLE
Remove Sabre's manual insertion-order iteration and unnecessary sorts

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

The manual insertion-order tracking was initially added in 02a1939 (gh-9012) during a complete rewrite of the scoring algorithms for Sabre to demonstrate that the new algorithm could be made fully RNG compatible.  It maintained the identical iteration order to the previous iteration once the front-layer data structure was swapped from being a raw list to hash-based.

The new data structure doesn't require the manual tracking to be reproducible as long as the iteration order is independent of the hash seed.  This swaps the relevant places over the `IndexMap` to be able to remove a lot of the manual tracking.  In casual testing, this didn't appear to have much effect on performance, but the resulting code is much simpler.

The sorts (and deliberate canonicalisation of the swaps) was necessary to match RNG compatibility in the pre-relative-score Sabre, but since the swaps are now iterated through in a deterministic order and guaranteed to be generated only once each (the previous version used a hashset to remove duplicates), neither step is necessary.

For a QV circuit of depth 5 at 1081 qubits on heavy hex, this is worth almost a 2x speedup in routing performance.

### Details and comments

I've still got some fun ideas for improving the handling of the extended set by a more data-structure-explicit segregation into layers within it.  I don't know at what point we should stop calling this "Sabre", because with all the changes we've made to the algorithm, the router only really bears a passing resemblance to the original paper.

### CODE REVIEW

**Comment by jakelishman** [comment]
Oops, I forgot to sort out the testing changes because of the new RNG patterns.  Not sure what's going on with the gate-direction stuff, but I suspect it's related to a swap `[x, y]` no longer being strictly ordered to have `x < y` in the output.

**Comment by jakelishman** [comment]
254fc51 fixes the RNG-related tests.  The remaining ones _could_ be fixed by changing the `CouplingMap`s in those tests to be bi-directional, but I think more properly the fix is #9561, which correctly registers `swap` as a symmetric gate in the `GateDirection` transpiler pass.

This PR now depends on #9561.

**Comment by alexanderivrii** [comment]
> I don't know at what point we should stop calling this "Sabre", because with all the changes we've made to the algorithm, the router only really bears a passing resemblance to the original paper.

Since you've made it so much more efficient, maybe it should be called `Lightsabre`?

**Comment by jakelishman** [comment]
Test non-determinism is fixed by #9568, but I'll need a new commit after that merges to update the chosen fixed layout in the currently flaky test to whatever that PR causes it to settle on.

File: releasenotes/notes/sabre-sort-rng-056f26f205e38bab.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+features:
+  - |
+    Internal tweaks to the routing algorithm in :class:`.SabreSwap`, used in
+    transpilation of non-dynamic circuits at all non-zero optimization levels,
+    have sped up routing for very large circuits.  As an example, the time to
+    route a depth-5 :class:`.QuantumVolume` circuit for a 1081-qubit heavy-hex
+    coupling map is approximately halved.
```
Comment (mtreinish) [inline releasenotes/notes/sabre-sort-rng-056f26f205e38bab.yaml:8]: I'd probably add an upgrade note here about results changing for fixed seed. While we don't explicitly guarantee seed for see compatibility between releases. Documenting this as a potential upgrade note seems like the right thing to do (I also think we've done this in previous releases when changing RNG behavior).

**Review by mtreinish — COMMENTED** [review]
Overall this LGTM, the code is pretty straightforward (or this could just be because we discussed it at the time as part of the #9012 review) and the speed boost is a big win (also shaking loose bugs in other parts of the code was great). Just one small inline comment about expanding the release notes, but once that's there I think this should be good to go.

File: releasenotes/notes/sabre-sort-rng-056f26f205e38bab.yaml (hunk 1 above)
Comment (jakelishman) [reply releasenotes/notes/sabre-sort-rng-056f26f205e38bab.yaml:8]: Huh, I actually thought I'd written that, but I guess not!  Done in 60ddde195.

**Review by mtreinish — APPROVED** [review]

**Pull request merged**